# TIL | 2025.03.29

## 📌 TIL 요약

### 알고리즘

-   정렬과 그리디 결정 알고리즘 10 ~ 12

### 소켓 적용

-   singleton으로 분리 Stompjs 객체

    -   처음에는 소켓 생성후 연결만 했었음

-   모르는거

    -   activate는 실제 연결 시도를 시작하는 메소드
        -   연결 완료/실패 시점을 알려면 onConnect/onStompError 핸들러를 통해서 연결 시작이 아닌 연결이 완료되거나 실패한 지점에 다음 동작을 수행하게 함

-   채승이 구현 방식
    -   소켓 instance 및 rtc , yjs 소켓 class 를 따로 작성
    -   소켓 매니저에서 한번에 관리

---

# 도서 : 좋은코드 나쁜코드

## 1장

-   코드 품질의 여섯 가지 핵심 요소
    1. 코드는 읽기 쉬워야 한다.
    2. 코드는 예측 가능해야 한다.
    3. 코드를 오용하기 어렵게 만들라.
    4. 코드를 모듈화하라
    5. 코드를 재사용 가능하고 일반화할 수 있게 작성하라.
    6. 테스트가 용이한 코드를 작성하고 제대로 테스트하라

### 1장 요약

> 좋은 소프트웨어를 만들려면 고품질 코드를 작성해야 한다.
> 실제 서비스 환경에서 실행되는 소프트웨어가 되기 전에는 코드는 일반적으로 여러 단계의 검사와 테스트를 통과애햐 한다.(때로는 수동 때로는 자동화를 통해서)
> 버그나 제대로 동작하지 않는 기능이 사용자에게 제공되거나 비즈니스에 중요한 시스템에서 실행되는 것을 이러한 검사를 통해 막을 수 있다.
> 테스트는 코드를 작성하는 모든 단계에서 고려하는 것이 좋아. 코드를 다 작성하고 난 후에 고려하는 것이 아니다.
> 고품질 코드를 작성하면 처음에는 시간이 오래 걸리지만, 중장기적으로는 개발 시간이 단축되는 경우가 많다.

## 추상화 계층

-   코드를 구성하는 방법은 코드 품질의 기본적인 측면 중 하나이며, 코드를 잘 구성한다는 것은 간결한 추상화 계층읋 만드는 것으로 귀결될 때가 많다.

### 2.2 왜 추상화 계층을 만드는가

-   코드 작성은 복잡한 문제를 계속해서 더 작은 문제로 세분화하는 작업이다.
    -   문제가 엄청나게 복잡할지라도 하위 문제들을 식별하고 올바른 추상화 계층을 만듦으로써 복잡한 문제를 쉽게 다룰 수 있다.
-   2.2.1 추상화 계층 및 코드 품질의 핵심 요소
    -   깨끗하고 뚜렷한 추상화 계층을 구축하면 코드 품질의 네 가지 핵심 요소를 만족 가능
        -   가독성, 모듈화, 재사용성 및 일반화서, 테스트 용이성

### 키워드

-   응집력
-   관심사의 분리
-   추상화 계층
-   팩토리 함수
-   인터페이스

### 1차 소켓 수정본

```ts
import { create } from 'zustand';
import { StompSocket, stompSocket } from '@/lib/socket';
import * as StompJs from '@stomp/stompjs';

interface SocketStore {
    socket: StompJs.Client | null;
    isStompSocketIsReady: boolean;
    connect: () => void;
    discconect: () => void;
    subscribe: (topic: string, callback: (message: any) => void) => void;
    publish: (topic: string, message?: string | undefined) => void;
}

export const useSocketStore = create<SocketStore>()((set, get) => ({
    socket: null,
    isStompSocketIsReady: false,
    connect: async () => {
        try {
            const stompClient = await StompSocket.getInstance().initializeSocket();
            console.log('store에서 stompclient', stompClient);
            set({ socket: stompClient, isStompSocketIsReady: true });
        } catch (e) {
            console.log(e);
        }
    },
    discconect: () => {
        const { socket, isStompSocketIsReady } = get();
        if (!socket || !isStompSocketIsReady!) return;
        socket.deactivate();
    },
    subscribe: (topic, callback) => {
        const { socket, isStompSocketIsReady } = get();
        if (!socket || !isStompSocketIsReady) return;
        socket.subscribe(topic, callback);
    },
    publish: (topic: string, message: string | undefined) => {
        const { socket, isStompSocketIsReady } = get();
        if (!socket || !isStompSocketIsReady) return;
        socket.publish({
            destination: topic,
            body: message ? message : undefined,
        });
    },
}));
```

```ts
import SockJS from 'sockjs-client';
import * as StompJs from '@stomp/stompjs';

const SERVER_URL = process.env.NEXT_PUBLIC_SERVER_DEV_URL;

export class StompSocket {
    private static instance: StompSocket;

    private constructor() {}

    public static getInstance() {
        if (!StompSocket.instance) {
            StompSocket.instance = new StompSocket();
        }
        return StompSocket.instance;
    }

    public initializeSocket() {
        return new Promise<StompJs.Client>((resolve, reject) => {
            const socket = new StompJs.Client({
                webSocketFactory: () => {
                    return new SockJS(`${SERVER_URL}/api/ws`);
                },
            });
            socket.onConnect = () => {
                console.log('socket 연결 성공');
                resolve(socket);
            };
            socket.onStompError = () => {
                console.log('socket 연결 실패');
            };
            socket.activate();
        });
    }
}

export const stompSocket = StompSocket.getInstance();
```

### 1차 구현 후 회고 및 수정할 부분 조사

-

###

```ts
import SockJS from 'sockjs-client';
import * as StompJs from '@stomp/stompjs';

const SERVER_URL = process.env.NEXT_PUBLIC_SERVER_DEV_URL;

export class StompSocket {
    private static instance: StompSocket;
    private client: StompJs.Client | null = null;

    private constructor() {}

    public static getInstance() {
        if (!StompSocket.instance) {
            StompSocket.instance = new StompSocket();
        }
        return StompSocket.instance;
    }

    public connect() {
        return new Promise<boolean>((resolve, reject) => {
            this.client = new StompJs.Client({
                webSocketFactory: () => {
                    return new SockJS(`${SERVER_URL}/api/ws`);
                },
            });
            this.client.onConnect = () => {
                console.log('socket 연결 성공');
                resolve(true);
            };
            this.client.onStompError = () => {
                console.log('socket 연결 실패');
                reject(false);
            };
            this.client.activate();
        });
    }

    public deactivate() {
        if (!this.checkIsConnected()) return;

        this.client?.deactivate();
    }

    public subscribe(topic: string, callback: (message: StompJs.IFrame) => void) {
        if (!this.checkIsConnected()) return;

        this.client?.subscribe(topic, callback);
    }

    public publish(topic: string, message?: string) {
        if (!this.checkIsConnected()) return;
        this.client?.publish({
            destination: topic,
            body: message ? message : undefined,
        });
    }

    private checkIsConnected() {
        if (!this.client || !this.client?.connected) {
            console.log('소컷 연결이 안됬습니다.');
            return false;
        }
        return true;
    }
}

export const stompSocket = StompSocket.getInstance();
```

## throw new Error(error) vs throw error

-   catch 에서 다시 throw 할 때
    -   try 에서 던지는 경우
        -   특정 로직에 따라 의도한 에러가 발생했을 경우 에러를 의도적으로 발생시키기 위해서
    -   catch에서 던지는 경우
        -   catch 블록에서 우리가 예상했던 특정 타입의 에러만 처리하고 다른 타입의 에러는 상위 호출자에게 전파하고 싶을 때 (= 에러 다시 던지기 )
        -   에러가 발생했을 때 그 에러의 메시지나 타입이 부적절하거나 충분한 정보를 제공하지 않는다고 판단되면, 새로운 에러 객체를 생성하여 그 정보를 포함하고 throw 하는 경우
        -   에러가 발생했을 때 특정 로직을 수행한 후에도 같은 에러를 계속 전파하고 싶을 때
-   catch 안에서 에러를 던질 때 2가지의 경우에 차이점

```ts
catch (error) {
    throw new Error(error);
}

catch (error) {
    throw error;
}
```

### throw new Error(error);

-   이 경우에는 새로운 Error 객체를 생성하여 던집니다.
-   error가 문자열이거나 toString 메서드를 가지고 있는 객체라면, 그 값을 메시지로 사용하여 새로운 Error 객체를 생성합니다.
    > 그렇지만 이 방법은 Error 객체를 새로 생성하므로 원래의 error 로 넘어온 에러 객체의 일부 정보를 잃을 수 있습니다

### throw error;

-   이 경우에는 error 객체 자체를 던집니다.
-   이것은 catch 블록에서 받은 원래의 에러 객체를 그대로 다시 던지는 것이므로, 원래의 에러 객체에 있던 모든 정보와 속성을 유지합니다.
    > 그렇지만 new Error(error) 형태로 에러를 래핑하고 던지면, 원래의 CustomError 객체의 code 속성과 같은 추가 정보는 사라지게 됩니다

### 참조

-   https://velog.io/@minh0518/throw-new-Errorerror-vs-throw-error
